File: intellij/src/main/kotlin/dev/nylo/plugin/localizations/service/LocalizationService.kt
package dev.nylo.plugin.localizations.service

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import dev.nylo.plugin.localizations.compare.LocaleComparator
import dev.nylo.plugin.localizations.edit.LangFileWriter
import dev.nylo.plugin.localizations.json.LangJson
import dev.nylo.plugin.localizations.model.LocalizationReport
import dev.nylo.plugin.localizations.scan.BaselineResolver
import dev.nylo.plugin.localizations.scan.LangFileScanner
import dev.nylo.plugin.state.NyloPluginState
import java.io.File

/**
 * Owns the localization data for a project so the tool window's tabs stay consistent.
 *
 * Reads + flattens every `lang/<code>.json` off the EDT, compares them with [LocaleComparator] (honoring the
 * persisted baseline + same-as-base toggle from [NyloPluginState]), caches the [report] and the
 * in-memory value maps, and publishes [LocalizationDataListener] on the EDT. Baseline/filter changes and
 * inline edits recompute from the in-memory maps without re-reading disk.
 */
@Service(Service.Level.PROJECT)
class LocalizationService(private val project: Project) {

    @Volatile
    var report: LocalizationReport = LocalizationReport.EMPTY
        private set

    @Volatile
    private var values: Map<String, LinkedHashMap<String, String>> = emptyMap()

    @Volatile
    private var files: Map<String, File> = emptyMap()

    @Volatile
    private var parseErrors: Map<String, String> = emptyMap()

    val localeCodes: List<String> get() = (values.keys + parseErrors.keys).sorted()

    fun valuesFor(locale: String): Map<String, String> = values[locale].orEmpty()

    fun fileFor(locale: String): File? = files[locale]

    /** Whether the project actually has a `lang/` directory (vs. just not being scanned yet). */
    fun hasLangDir(): Boolean = LangFileScanner.langDir(project)?.isDirectory == true

    /** Reloads every lang file off the EDT, recomputes the report, then publishes on the EDT. */
    fun refreshAsync() {
        val app = ApplicationManager.getApplication()
        app.executeOnPooledThread {
            if (project.isDisposed) return@executeOnPooledThread
            val newValues = LinkedHashMap<String, LinkedHashMap<String, String>>()
            val newFiles = LinkedHashMap<String, File>()
            val newErrors = LinkedHashMap<String, String>()
            for (lf in LangFileScanner.scan(project)) {
                newFiles[lf.code] = lf.file
                try {
                    newValues[lf.code] = LangJson.readFlattened(lf.file)
                } catch (e: Exception) {
                    newErrors[lf.code] = e.message ?: "Invalid JSON"
                }
            }
            values = newValues
            files = newFiles
            parseErrors = newErrors
            publish(compute())
        }
    }

    /** Recompares the in-memory maps with the current baseline/flag settings (no disk I/O). */
    fun recomputeAsync() {
        ApplicationManager.getApplication().executeOnPooledThread {
            if (!project.isDisposed) publish(compute())
        }
    }

    private fun compute(): LocalizationReport {
        val state = NyloPluginState.getInstance(project)
        val codes = (values.keys + parseErrors.keys).sorted()
        val baseline = BaselineResolver.resolve(codes, state.localizationsBaseline, envDefaultLocale()) ?: ""
        return LocaleComparator.compare(
            baseline = baseline,
            localeValues = values,
            parseErrors = parseErrors,
            flagSameAsBase = state.localizationsShowSameAsBase,
        )
    }

    private fun publish(newReport: LocalizationReport) {
        report = newReport
        ApplicationManager.getApplication().invokeLater({
            if (!project.isDisposed) {
                project.messageBus.syncPublisher(LocalizationDataListener.TOPIC).localizationDataChanged()
            }
        }, ModalityState.nonModal())
    }

    private fun envDefaultLocale(): String? {
        val basePath = project.basePath ?: return null
        return BaselineResolver.readEnvDefaultLocale(File(basePath, ".env"))
    }

    /** Persists a new baseline override (null clears it back to the resolved default) and recomputes. */
    fun setBaselineOverride(code: String?) {
        NyloPluginState.getInstance(project).localizationsBaseline = code
        recomputeAsync()
    }

    /** Persists the same-as-base toggle and recomputes (it changes which issues are produced). */
    fun setFlagSameAsBase(enabled: Boolean) {
        NyloPluginState.getInstance(project).localizationsShowSameAsBase = enabled
        recomputeAsync()
    }

    /**
     * EDT-only inline edit: writes [value] for [key] into [locale]'s file, updates the in-memory map so
     * the UI reflects it immediately, then recomputes. Returns null on success, else an error message.
     */
    fun setValue(locale: String, key: String, value: String): String? {
        val file = files[locale] ?: return "No lang file for $locale"
        // No-change commits (double-click a cell, press Enter) must not rewrite the file: the write
        // re-serializes the whole JSON (reformatting hand-formatted files) and triggers a refresh cycle.
        if (values[locale]?.get(key) == value) return null
        val error = LangFileWriter.setValue(project, file, key, value)
        if (error != null) return error
        values[locale]?.let { existing ->
            val copy = LinkedHashMap(existing).apply { put(key, value) }
            values = HashMap(values).apply { put(locale, copy) }
        }
        recomputeAsync()
        return null
    }

    companion object {
        const val TOOL_WINDOW_ID = "Nylo Localizations"

        fun getInstance(project: Project): LocalizationService = project.service()
    }
}

File: vscode/src/host/localizations/localizationService.ts
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';
import { readEnvDefaultLocale, resolveBaseline } from '../../core/localizations/baselineResolver';
import { LANG_DIR, LangFile, scanLangFiles } from '../../core/localizations/langFileScanner';
import { parseFlattened } from '../../core/localizations/langJson';
import { compareLocales } from '../../core/localizations/localeComparator';
import { EMPTY_REPORT, LocalizationReport } from '../../core/localizations/model';
import { writeValue } from './langFileWriter';

const KEY_BASELINE = 'nylo.loc.baseline';
const KEY_SAME = 'nylo.loc.sameAsBase';

/**
 * Port of `dev.nylo.plugin.localizations.service.LocalizationService`: reads + flattens the lang files,
 * caches the per-locale value maps, recomputes the report on baseline/flag/edit changes, and publishes
 * via {@link onDidChange}.
 */
export class LocalizationService implements vscode.Disposable {
  private readonly emitter = new vscode.EventEmitter<void>();
  readonly onDidChange = this.emitter.event;

  private folder: vscode.WorkspaceFolder | undefined;
  private langFiles: LangFile[] = [];
  private values = new Map<string, Map<string, string>>();
  private parseErrors = new Map<string, string>();
  private report: LocalizationReport = EMPTY_REPORT;
  private baselineOverride: string | null;
  private flagSameAsBase: boolean;

  constructor(private readonly context: vscode.ExtensionContext) {
    this.baselineOverride = context.workspaceState.get<string | null>(KEY_BASELINE, null);
    this.flagSameAsBase = context.workspaceState.get<boolean>(KEY_SAME, false);
  }

  setFolder(folder: vscode.WorkspaceFolder | undefined): void {
    this.folder = folder;
    this.refresh();
  }

  getReport(): LocalizationReport {
    return this.report;
  }

  getValues(): Map<string, Map<string, string>> {
    return this.values;
  }

  getLocaleCodes(): string[] {
    return this.langFiles.map((f) => f.code);
  }

  getFlagSameAsBase(): boolean {
    return this.flagSameAsBase;
  }

  fileFor(locale: string): string | undefined {
    return this.langFiles.find((f) => f.code === locale)?.filePath;
  }

  /** Whether the project actually has a lang/ directory (vs. just not being scanned yet). */
  hasLangDir(): boolean {
    if (!this.folder) {
      return false;
    }
    try {
      return fs.statSync(path.join(this.folder.uri.fsPath, LANG_DIR)).isDirectory();
    } catch {
      return false;
    }
  }

  /** Re-reads all lang files from disk and recomputes. */
  refresh(): void {
    this.values = new Map();
    this.parseErrors = new Map();
    if (!this.folder) {
      this.langFiles = [];
      this.report = EMPTY_REPORT;
      this.emitter.fire();
      return;
    }
    this.langFiles = scanLangFiles(this.folder.uri.fsPath);
    for (const lang of this.langFiles) {
      try {
        this.values.set(lang.code, parseFlattened(fs.readFileSync(lang.filePath, 'utf8')));
      } catch (e) {
        this.parseErrors.set(lang.code, (e as Error).message);
      }
    }
    this.recompute();
  }

  /** Recomputes the report from the in-memory maps (no disk read) — for baseline/flag/edit changes. */
  recompute(): void {
    const codes = this.langFiles.map((f) => f.code);
    const baseline = resolveBaseline(codes, this.baselineOverride, this.readEnvDefault());
    this.report = baseline
      ? compareLocales(baseline, this.values, this.parseErrors, this.flagSameAsBase)
      : EMPTY_REPORT;
    this.emitter.fire();
  }

  setBaselineOverride(code: string | null): void {
    this.baselineOverride = code;
    this.context.workspaceState.update(KEY_BASELINE, code);
    this.recompute();
  }

  toggleSameAsBase(): void {
    this.flagSameAsBase = !this.flagSameAsBase;
    this.context.workspaceState.update(KEY_SAME, this.flagSameAsBase);
    this.recompute();
  }

  async setValue(locale: string, key: string, value: string): Promise<string | null> {
    const file = this.fileFor(locale);
    if (!file) {
      return `No lang file for ${locale}`;
    }
    // No-change commits must not rewrite the file: the write re-serialises the whole JSON
    // (reformatting hand-formatted files) and triggers a refresh cycle.
    if (this.values.get(locale)?.get(key) === value) {
      return null;
    }
    const error = await writeValue(file, key, value);
    if (error) {
      return error;
    }
    // Only update a locale that already has a parsed value map — never fabricate one for a locale
    // with none (e.g. a parse-error locale), which would make the comparator treat it as a 1-key file.
    this.values.get(locale)?.set(key, value);
    this.recompute();
    return null;
  }

  private readEnvDefault(): string | null {
    if (!this.folder) {
      return null;
    }
    try {
      return readEnvDefaultLocale(fs.readFileSync(path.join(this.folder.uri.fsPath, '.env'), 'utf8'));
    } catch {
      return null;
    }
  }

  dispose(): void {
    this.emitter.dispose();
  }
}
